homeassistant/components/vallox/fan.py

```diff
@@ -0,0 +1,167 @@
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def available(self):
+        """Return if state is known."""
+        return self._available
+
+    @property
+    def is_on(self):
+        """Return if device is on."""
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Return device specific state attributes."""
+        return {
+            ATTR_PROFILE_FAN_SPEED_HOME['description']: self._fan_speed_home,
+            ATTR_PROFILE_FAN_SPEED_AWAY['description']: self._fan_speed_away,
+            ATTR_PROFILE_FAN_SPEED_BOOST['description']: self._fan_speed_boost,
+        }
+
+    async def async_added_to_hass(self):
+        """Call to update."""
+        async_dispatcher_connect(self.hass, SIGNAL_VALLOX_STATE_UPDATE,
+                                 self._update_callback)
+
+    @callback
+    def _update_callback(self):
+        """Call update method."""
+        self.async_schedule_update_ha_state(True)
+
+    async def async_update(self):
+        """Fetch state from the device."""
+        try:
+            # Fetch if the whole device is in regular operation state.
+            mode = self._state_proxy.fetch_metric(METRIC_KEY_MODE)
+            if mode == 0:
+                self._state = True
+            else:
+                self._state = None
+
+            # Fetch the profile fan speeds.
+            self._fan_speed_home = int(self._state_proxy.fetch_metric(
+                ATTR_PROFILE_FAN_SPEED_HOME['metric_key']))
+            self._fan_speed_away = int(self._state_proxy.fetch_metric(
+                ATTR_PROFILE_FAN_SPEED_AWAY['metric_key']))
+            self._fan_speed_boost = int(self._state_proxy.fetch_metric(
+                ATTR_PROFILE_FAN_SPEED_BOOST['metric_key']))
+
+            self._available = True
+
+        except (IOError, KeyError) as err:
+            self._available = False
+            _LOGGER.error("Error updating fan. %s", err)
+
+    async def async_turn_on(self, speed: str = None, **kwargs) -> None:
+        """Turn the device on."""
+        _LOGGER.debug("Turn on: %s", speed)
+
+        # Case speed == None equals the GUI toggle switch being activated.
+        if speed is None:
+            # Switch to regular operation only if device is currently in other
+            # mode.
+            if self._state is None:
+                try:
+                    self._client.set_values({METRIC_KEY_MODE: 0})
+
+                    # This state change affects other entities like sensors.
+                    # Force an immediate update that can be observed by all
+                    # parties involved.
+                    await self._state_proxy.async_update(None)
+
+                except IOError as io_err:
+                    self._available = False
+                    _LOGGER.error("Error turning on: %s", io_err)
+            else:
+                _LOGGER.error("Already on.")
```

Comment:
If we have the wrong state info we won't be able to change state. If that's a possibility, we try to avoid guards like this.

Reply:
I put it there to catch a call to the service `fan/turn_on`, where tests showed it executes even if the fan is already on.
This could for example happen if a user-automation script is erroneous.

Reply:
Would it be bad for the device to call turn on if it's already on?

Reply:
I don't know. There's no documentation that would give it away, so I did not want to take chances and cause side effects in the device when we can prevent it on the higher level.

Reply:
I think it's better to put protection close to the device if the device requires protection.

Reply:
The problem is we cannot find out if the device firmware is fine with mutliple calls or if there would be unforeseen consequences if it would be spammed with those commands due to a user script gone rogue (making up a "dramatic" case here).

In the end it is just an if/else on a boolean to be on the safe side here. IMO, that's an acceptable price to pay for a little more safety?